src/python/pants/rules/core/list_target_types.py

```diff
@@ -0,0 +1,107 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from typing import Type
+
+from pants.engine.console import Console
+from pants.engine.goal import Goal, GoalSubsystem, LineOriented
+from pants.engine.rules import UnionMembership, goal_rule
+from pants.engine.target import BoolField, Field, RegisteredTargetTypes, Target
+from pants.util.objects import get_first_line_of_docstring
+
+
+class TargetTypesOptions(LineOriented, GoalSubsystem):
+    """List all the registered target types, including custom plugin types."""
+
+    # TODO: consider renaming this to to `target-types` or even a more general goal like
+    #  `symbols --type=targets`.
+    name = "targets2"
+
+    @classmethod
+    def register_options(cls, register):
+        super().register_options(register)
+        register(
+            "--details",
+            type=str,
+            metavar="target_type",
+            help="List all of the target type's registered fields.",
+        )
+
+
+class TargetTypes(Goal):
+    subsystem_cls = TargetTypesOptions
+
+
+def abbreviated_target_information(target_type: Type[Target], console: Console) -> str:
+    """Return a single line description of the target type."""
+    formatted_alias = console.cyan(f"{target_type.alias:>30}:")
+    description = get_first_line_of_docstring(target_type) or "<no description>"
+    return f"{formatted_alias} {description}"
+
+
+def verbose_target_information(
+    target_type: Type[Target], console: Console, union_membership: UnionMembership
+) -> str:
+    """Return a multiline description of the target type and all of its fields."""
+    # TODO: show the full docstring for both the target type and each field. We first need to
+    #  process them, e.g. removing all blank lines and possibly stripping type hints? See
+    #  build_dictionary_info_extractor.py for inspiration.
+    target_type_description = get_first_line_of_docstring(target_type)
+    if target_type_description:
+        target_type_description = console.blue(target_type_description)
+
+    def format_field(field: Type[Field]) -> str:
+        # TODO: possibly put type information in this output. We would extract it from the type
+        #  hints, but it might be weird to handle Optional.
+        field_alias_text = f"  {field.alias} = ...,"
+        field_alias = console.cyan(f"{field_alias_text:<30}")
+        description = get_first_line_of_docstring(field) or "<no description>"
+        # TODO: should we elevate `default` so that every Field has it, whereas now only bool
+        #  fields have it? V1 `targets` renders a default value, which is helpful.
```

Comment:
I don't know how I feel about this. 

It means that we likely would no longer want to make this be abstract:

https://github.com/pantsbuild/pants/blob/2d6853ce915048504cc9ed31222aedd69b7bf588/src/python/pants/engine/target.py#L95-L104

Naturally, we could provide an implementation `return raw_value if raw_value is not None else self.default`. But, I like requiring subclasses to have to implement the method (e.g. so that they know to add runtime type checking) and to avoid needing to call `super().hydrate`.

I suppose that we could keep `.hdyrated()` abstract and still define the class property `default`. But that means that we have no guarantees people are actually using that default.

FTR, default here means "what do you do when `raw_value is None`.

--

After typing that all out, I'm less opposed to this than I was before. Particularly because every field author should be using something like `StringField` rather than directly `PrimitiveField`.

Thoughts?

Reply:
I made this change in a followup PR and feel really good about it now. It simplifies `Sources` to use `default` instead of `default_globs`.